In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("vehicles_messy.csv", low_memory=False)

In [3]:
# how many columns?
len(df.columns)

83

In [4]:
# no of values
df.size

3140969

In [5]:
# no of rows
len(df)

37843

In [6]:
df.shape   # dimension

(37843, 83)

In [7]:
df.head()

barrels08  barrelsA08  charge120  charge240  city08  city08U  cityA08  \
0  15.695714         0.0        0.0        0.0      19      0.0        0   
1  29.964545         0.0        0.0        0.0       9      0.0        0   
2  12.207778         0.0        0.0        0.0      23      0.0        0   
3  29.964545         0.0        0.0        0.0      10      0.0        0   
4  17.347895         0.0        0.0        0.0      17      0.0        0   

   cityA08U  cityCD  cityE  ...  mfrCode  c240Dscr  charge240b  c240bDscr  \
0       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
1       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
2       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
3       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   
4       0.0     0.0    0.0  ...      NaN       NaN         0.0        NaN   

                      createdOn                    modifiedOn  startStop  \
0  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
1  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
2  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
3  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
4  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   

   phevCity  phevHwy  phevComb  
0         0        0         0  
1         0        0         0  
2         0        0         0  
3         0        0         0  
4         0        0         0  

[5 rows x 83 columns]

### Data Cleaning Steps

    1. Handle missing values
    2. Look at information contained in a column (Cardinality / Variance)
    3. Detect Outliers
    4. Convert types
    5. String operations (Parsing)
    6. Treat duplicates

### 1.Handle missing values

In [8]:
df.isna() # for every missing value is True

barrels08  barrelsA08  charge120  charge240  city08  city08U  cityA08  \
0          False       False      False      False   False    False    False   
1          False       False      False      False   False    False    False   
2          False       False      False      False   False    False    False   
3          False       False      False      False   False    False    False   
4          False       False      False      False   False    False    False   
...          ...         ...        ...        ...     ...      ...      ...   
37838      False       False      False      False   False    False    False   
37839      False       False      False      False   False    False    False   
37840      False       False      False      False   False    False    False   
37841      False       False      False      False   False    False    False   
37842      False       False      False      False   False    False    False   

       cityA08U  cityCD  cityE  ...  mfrCode  c240Dscr  charge240b  c240bDscr  \
0         False   False  False  ...     True      True       False       True   
1         False   False  False  ...     True      True       False       True   
2         False   False  False  ...     True      True       False       True   
3         False   False  False  ...     True      True       False       True   
4         False   False  False  ...     True      True       False       True   
...         ...     ...    ...  ...      ...       ...         ...        ...   
37838     False   False  False  ...     True      True       False       True   
37839     False   False  False  ...     True      True       False       True   
37840     False   False  False  ...     True      True       False       True   
37841     False   False  False  ...     True      True       False       True   
37842     False   False  False  ...     True      True       False       True   

       createdOn  modifiedOn  startStop  phevCity  phevHwy  phevComb  
0          False       False       True     False    False     False  
1          False       False       True     False    False     False  
2          False       False       True     False    False     False  
3          False       False       True     False    False     False  
4          False       False       True     False    False     False  
...          ...         ...        ...       ...      ...       ...  
37838      False       False       True     False    False     False  
37839      False       False       True     False    False     False  
37840      False       False       True     False    False     False  
37841      False       False       True     False    False     False  
37842      False       False       True     False    False     False  

[37843 rows x 83 columns]

In [9]:
# missing values for each column (vertically; it can be modified for rows, see pandas website)
df.isna().sum()

barrels08         0
barrelsA08        0
charge120         0
charge240         0
city08            0
              ...  
modifiedOn        0
startStop     31705
phevCity          0
phevHwy           0
phevComb          0
Length: 83, dtype: int64

In [10]:
type(df.isna().sum())

pandas.core.series.Series

In [11]:
# only show the columns that have missing values
missing_values = df.isna().sum()

In [12]:
missing_values[missing_values.gt(0)] # it'll show the columns with values greater than 0

cylinders       123
displ           120
drive          1189
eng_dscr      15403
trany            11
guzzler       35562
trans_dscr    22796
tCharger      32657
sCharger      37177
atvType       34771
fuelType2     36435
rangeA        36440
evMotor       37281
mfrCode       30818
c240Dscr      37806
c240bDscr     37807
startStop     31705
dtype: int64

In [13]:
missing_values[missing_values > 0]

cylinders       123
displ           120
drive          1189
eng_dscr      15403
trany            11
guzzler       35562
trans_dscr    22796
tCharger      32657
sCharger      37177
atvType       34771
fuelType2     36435
rangeA        36440
evMotor       37281
mfrCode       30818
c240Dscr      37806
c240bDscr     37807
startStop     31705
dtype: int64

In [14]:
# relative
missing_values[missing_values > 0] / len(df) # computing the ratio of missing values

cylinders     0.003250
displ         0.003171
drive         0.031419
eng_dscr      0.407024
trany         0.000291
guzzler       0.939725
trans_dscr    0.602384
tCharger      0.862960
sCharger      0.982401
atvType       0.918823
fuelType2     0.962794
rangeA        0.962926
evMotor       0.985149
mfrCode       0.814365
c240Dscr      0.999022
c240bDscr     0.999049
startStop     0.837804
dtype: float64

In [15]:
# this is how you drop columns, but first one must know WHY
df.drop(columns=["evMotor", "c240bDscr"]).head()

barrels08  barrelsA08  charge120  charge240  city08  city08U  cityA08  \
0  15.695714         0.0        0.0        0.0      19      0.0        0   
1  29.964545         0.0        0.0        0.0       9      0.0        0   
2  12.207778         0.0        0.0        0.0      23      0.0        0   
3  29.964545         0.0        0.0        0.0      10      0.0        0   
4  17.347895         0.0        0.0        0.0      17      0.0        0   

   cityA08U  cityCD  cityE  ...  rangeA  mfrCode  c240Dscr  charge240b  \
0       0.0     0.0    0.0  ...     NaN      NaN       NaN         0.0   
1       0.0     0.0    0.0  ...     NaN      NaN       NaN         0.0   
2       0.0     0.0    0.0  ...     NaN      NaN       NaN         0.0   
3       0.0     0.0    0.0  ...     NaN      NaN       NaN         0.0   
4       0.0     0.0    0.0  ...     NaN      NaN       NaN         0.0   

                      createdOn                    modifiedOn  startStop  \
0  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
1  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
2  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
3  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   
4  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013        NaN   

   phevCity  phevHwy  phevComb  
0         0        0         0  
1         0        0         0  
2         0        0         0  
3         0        0         0  
4         0        0         0  

[5 rows x 81 columns]

In [16]:
# calculate the ratio of missing values for each column
missing_value_ratios = missing_values[missing_values > 0] / len(df)
missing_value_ratios

cylinders     0.003250
displ         0.003171
drive         0.031419
eng_dscr      0.407024
trany         0.000291
guzzler       0.939725
trans_dscr    0.602384
tCharger      0.862960
sCharger      0.982401
atvType       0.918823
fuelType2     0.962794
rangeA        0.962926
evMotor       0.985149
mfrCode       0.814365
c240Dscr      0.999022
c240bDscr     0.999049
startStop     0.837804
dtype: float64

In [17]:
# retrieve the column names for those columns that have more than 50% missing values
missing_value_ratios[missing_value_ratios > .5].index

Index(['guzzler', 'trans_dscr', 'tCharger', 'sCharger', 'atvType', 'fuelType2',
       'rangeA', 'evMotor', 'mfrCode', 'c240Dscr', 'c240bDscr', 'startStop'],
      dtype='object')

In [18]:
# and finally, drop them, so we reassign with "df = "
df = df.drop(columns=missing_value_ratios[missing_value_ratios > .5].index)

In [19]:
(df.isna().sum())[df.isna().sum() > 0] # shows the columns filtered

cylinders      123
displ          120
drive         1189
eng_dscr     15403
trany           11
dtype: int64

In [21]:
def cols_missing_value(df):
    return (df.isna().sum())[df.isna().sum() > 0]

In [22]:
cols_missing_value(df)

cylinders      123
displ          120
drive         1189
eng_dscr     15403
trany           11
dtype: int64

In [23]:
# get a subset of columns used for investigation
relevant_columns = ["year", "make", "model", "trany", "drive", "fuelType", "cylinders", "displ"]
df.loc[df["cylinders"].isna(), relevant_columns].head(30)

year        make                            model           trany  \
7138   2000      Nissan                         Altra EV             NaN   
7139   2000      Toyota                          RAV4 EV             NaN   
8143   2001      Toyota                          RAV4 EV             NaN   
8144   2001        Ford                            Th!nk             NaN   
8146   2001        Ford           Explorer USPS Electric             NaN   
8147   2001      Nissan                       Hyper-Mini             NaN   
9212   2002      Toyota                          RAV4 EV             NaN   
9213   2002        Ford           Explorer USPS Electric             NaN   
10329  2003      Toyota                          RAV4 EV             NaN   
21413  1985      Subaru                         RX Turbo    Manual 5-spd   
21414  1985      Subaru                         RX Turbo    Manual 5-spd   
21506  1986       Mazda                             RX-7    Manual 5-spd   
23026  2001        Ford                Ranger Pickup 2WD  Automatic (A1)   
23027  2000        Ford                Ranger Pickup 2WD  Automatic (A1)   
23028  2000        Ford                Ranger Pickup 2WD  Automatic (A1)   
23029  1999         GMC                              EV1  Automatic (A1)   
23030  1999         GMC                              EV1  Automatic (A1)   
23032  1999       Honda                          EV Plus  Automatic (A1)   
23033  1999        Ford                Ranger Pickup 2WD  Automatic (A1)   
23034  1999        Ford                Ranger Pickup 2WD  Automatic (A1)   
23035  1999       Dodge        Caravan/Grand Caravan 2WD  Automatic (A1)   
23036  1999    Plymouth        Voyager/Grand Voyager 2WD  Automatic (A1)   
23037  1998       Honda                          EV Plus  Automatic (A1)   
23038  1998   Chevrolet                     S10 Electric  Automatic (A1)   
23039  1998   Chevrolet                     S10 Electric  Automatic (A1)   
23040  2008        MINI                            MiniE  Automatic (A1)   
23041  2011      Nissan                             Leaf  Automatic (A1)   
23137  2011       smart  fortwo electric drive cabriolet  Automatic (A1)   
23138  2011       smart      fortwo electric drive coupe  Automatic (A1)   
23807  2012  Mitsubishi                           i-MiEV  Automatic (A1)   

                   drive     fuelType  cylinders  displ  
7138                 NaN  Electricity        NaN    NaN  
7139       2-Wheel Drive  Electricity        NaN    NaN  
8143       2-Wheel Drive  Electricity        NaN    NaN  
8144                 NaN  Electricity        NaN    NaN  
8146       2-Wheel Drive  Electricity        NaN    NaN  
8147                 NaN  Electricity        NaN    NaN  
9212       2-Wheel Drive  Electricity        NaN    NaN  
9213       2-Wheel Drive  Electricity        NaN    NaN  
10329      2-Wheel Drive  Electricity        NaN    NaN  
21413      4-Wheel Drive      Regular        NaN    NaN  
21414      4-Wheel Drive      Regular        NaN    NaN  
21506   Rear-Wheel Drive      Regular        NaN    1.3  
23026      2-Wheel Drive  Electricity        NaN    NaN  
23027      2-Wheel Drive  Electricity        NaN    NaN  
23028      2-Wheel Drive  Electricity        NaN    NaN  
23029                NaN  Electricity        NaN    NaN  
23030                NaN  Electricity        NaN    NaN  
23032                NaN  Electricity        NaN    NaN  
23033      2-Wheel Drive  Electricity        NaN    NaN  
23034      2-Wheel Drive  Electricity        NaN    NaN  
23035      2-Wheel Drive  Electricity        NaN    NaN  
23036  Front-Wheel Drive  Electricity        NaN    NaN  
23037                NaN  Electricity        NaN    NaN  
23038      2-Wheel Drive  Electricity        NaN    NaN  
23039      2-Wheel Drive  Electricity        NaN    NaN  
23040                NaN  Electricity        NaN    NaN  
23041  Front-Wheel Drive  Electricity        NaN    NaN  
23137   Rear-Wheel Drive  

In [24]:
df.loc[df["cylinders"].isna(), "fuelType"].value_counts() # From the rows, 120 have fuelType=Electricity

Electricity    120
Regular          3
Name: fuelType, dtype: int64

Obviously, electric cars don't have cylinders. There are, however, 3 cases where "regular" fuelType has missing cylinders.

###### Note: During class, the below operations did not work. The code was later corrected  so that it can be seen how it was initially planned on running it:

Impute the missing values for electric cars with 0.

In [25]:
df.loc[(df["cylinders"].isna()) & (df["fuelType"] == "Electricity"), "cylinders"] = 0

In [26]:
cols_missing_value(df)  # Now, the missing 3 cases finally appear

cylinders        3
displ          120
drive         1189
eng_dscr     15403
trany           11
dtype: int64

In [27]:
df.loc[df["cylinders"].isna(), relevant_columns]

year    make     model         trany             drive fuelType  \
21413  1985  Subaru  RX Turbo  Manual 5-spd     4-Wheel Drive  Regular   
21414  1985  Subaru  RX Turbo  Manual 5-spd     4-Wheel Drive  Regular   
21506  1986   Mazda      RX-7  Manual 5-spd  Rear-Wheel Drive  Regular   

       cylinders  displ  
21413        NaN    NaN  
21414        NaN    NaN  
21506        NaN    1.3

In [28]:
df["cylinders"] = df["cylinders"].fillna(0)

During class, the operations above did not run. Instead, we found this workaround:

Impute all the missing values with 0. We still remember which 3 should not be 0.

In [29]:
df["cylinders"] = df["cylinders"].fillna(0)

In [30]:
cols_missing_value(df)

displ         120
drive        1189
eng_dscr    15403
trany          11
dtype: int64

In [31]:
df.loc[df["fuelType"] != "Electricity"].loc[df["cylinders"] == 0, relevant_columns]

year    make     model         trany             drive fuelType  \
21413  1985  Subaru  RX Turbo  Manual 5-spd     4-Wheel Drive  Regular   
21414  1985  Subaru  RX Turbo  Manual 5-spd     4-Wheel Drive  Regular   
21506  1986   Mazda      RX-7  Manual 5-spd  Rear-Wheel Drive  Regular   

       cylinders  displ  
21413        0.0    NaN  
21414        0.0    NaN  
21506        0.0    1.3

## 2. Look at information contained in a column (Cardinality / Variance)

#### Example: What if we had one column with only 1 value

In [32]:
len(df
 .assign(vehicle=True)["vehicle"]
 .value_counts())

1

This column would not give us any useful information as it only has 1 unique value.

Look at other numerical columns:

In [34]:
df[["barrels08"]].sort_values("barrels08")

barrels08
25835   0.060000
24539   0.060000
27174   0.060000
28455   0.060000
14370   0.066429
...          ...
31336  47.087143
7901   47.087143
21060  47.087143
33860  47.087143
32586  47.087143

[37843 rows x 1 columns]

In [35]:
df[["barrels08"]].describe() # Gives the 5 point summary of a column

barrels08
count  37843.000000
mean      17.532506
std        4.575950
min        0.060000
25%       14.330870
50%       17.347895
75%       20.600625
max       47.087143

If 90% of the rows in a column have the same value, we might as well drop it.

- 1. for a given column, identify the min value
- 2. for the same column, identify the 90th percentile
- 3. if min == 90th perc, drop the column

In [36]:
# 1
min_b = df["barrels08"].min()

# 2
n_perc = np.percentile(df["barrels08"], 90)

# 3
min_b == n_perc

False

In this case, we would not drop `barrels08`

In [37]:
df.columns

Index(['barrels08', 'barrelsA08', 'charge120', 'charge240', 'city08',
       'city08U', 'cityA08', 'cityA08U', 'cityCD', 'cityE', 'cityUF', 'co2',
       'co2A', 'co2TailpipeAGpm', 'co2TailpipeGpm', 'comb08', 'comb08U',
       'combA08', 'combA08U', 'combE', 'combinedCD', 'combinedUF', 'cylinders',
       'displ', 'drive', 'engId', 'eng_dscr', 'feScore', 'fuelCost08',
       'fuelCostA08', 'fuelType', 'fuelType1', 'ghgScore', 'ghgScoreA',
       'highway08', 'highway08U', 'highwayA08', 'highwayA08U', 'highwayCD',
       'highwayE', 'highwayUF', 'hlv', 'hpv', 'id', 'lv2', 'lv4', 'make',
       'model', 'mpgData', 'phevBlended', 'pv2', 'pv4', 'range', 'rangeCity',
       'rangeCityA', 'rangeHwy', 'rangeHwyA', 'trany', 'UCity', 'UCityA',
       'UHighway', 'UHighwayA', 'VClass', 'year', 'youSaveSpend', 'charge240b',
       'createdOn', 'modifiedOn', 'phevCity', 'phevHwy', 'phevComb'],
      dtype='object')

In [38]:
df.select_dtypes(include=[np.number]).columns # Give all columns w/numerical value

Index(['barrels08', 'barrelsA08', 'charge120', 'charge240', 'city08',
       'city08U', 'cityA08', 'cityA08U', 'cityCD', 'cityE', 'cityUF', 'co2',
       'co2A', 'co2TailpipeAGpm', 'co2TailpipeGpm', 'comb08', 'comb08U',
       'combA08', 'combA08U', 'combE', 'combinedCD', 'combinedUF', 'cylinders',
       'displ', 'engId', 'feScore', 'fuelCost08', 'fuelCostA08', 'ghgScore',
       'ghgScoreA', 'highway08', 'highway08U', 'highwayA08', 'highwayA08U',
       'highwayCD', 'highwayE', 'highwayUF', 'hlv', 'hpv', 'id', 'lv2', 'lv4',
       'pv2', 'pv4', 'range', 'rangeCity', 'rangeCityA', 'rangeHwy',
       'rangeHwyA', 'UCity', 'UCityA', 'UHighway', 'UHighwayA', 'year',
       'youSaveSpend', 'charge240b', 'phevCity', 'phevHwy', 'phevComb'],
      dtype='object')

In [39]:
# get a list of all columns that we would drop:
to_drop = []

for col in df.select_dtypes(include=[np.number]).columns:
    min_value = df[col].min()                     # "col" was "barrels08" above
    perc_value = np.percentile(df[col], 90)
    if min_value == perc_value:
        to_drop.append(col)

In [40]:
to_drop # a list with all columns w/relatively low cardinality

['barrelsA08',
 'charge120',
 'charge240',
 'cityA08',
 'cityA08U',
 'cityCD',
 'cityE',
 'cityUF',
 'co2A',
 'co2TailpipeAGpm',
 'combA08',
 'combA08U',
 'combE',
 'combinedCD',
 'combinedUF',
 'fuelCostA08',
 'ghgScoreA',
 'highwayA08',
 'highwayA08U',
 'highwayCD',
 'highwayE',
 'highwayUF',
 'range',
 'rangeCity',
 'rangeCityA',
 'rangeHwy',
 'rangeHwyA',
 'UCityA',
 'UHighwayA',
 'charge240b',
 'phevCity',
 'phevHwy',
 'phevComb']

Alright, then drop those columns (cuz they don't provide a lot of Info)

In [41]:
df = df.drop(columns=to_drop)

### 3. Detect Outliers

Using Interquartile Range (IQR)

- 1. For a given column, calculate the IQR (75th perc - 25th perc)
- 2. determine a tolerance factor
- 3. determine the lower and upper threshold
- 4. label all values that are beyond the thresholds as outliers



In [42]:
# 1
barrels_desc = df[["barrels08"]].describe().transpose() # making it horizontal
barrels_desc = barrels_desc.assign(IQR=barrels_desc["75%"] - barrels_desc["25%"]) #new column

In [43]:
barrels_desc

count       mean      std   min       25%        50%        75%  \
barrels08  37843.0  17.532506  4.57595  0.06  14.33087  17.347895  20.600625   

                 max       IQR  
barrels08  47.087143  6.269755

In [44]:
barrels_desc.loc["barrels08","25%"] # the specific value of the chart

14.330869565217393

In [45]:
# 2 & 3
tol_factor = 3      # arbitrarily chosen, if it's too tolerant, one can lower it
barrels_lower_threshold = barrels_desc.loc["barrels08","25%"] - tol_factor * barrels_desc.loc["barrels08","IQR"]
barrels_upper_threshold = barrels_desc.loc["barrels08","75%"] + tol_factor * barrels_desc.loc["barrels08","IQR"]

In [46]:
barrels_lower_threshold

-4.47839673913043

In [47]:
barrels_upper_threshold

39.40989130434782

In this scenario, we would label all rows as 'outliers' if their value for barrels08 is beyond the thresholds. That is, if barrels08 is larger than 39.41

For all numerical columns:

In [48]:
descr = df["barrels08"].describe().transpose()

In [49]:
df[["barrels08"]].describe().transpose() # Option A to correctly transpose

count       mean      std   min       25%        50%        75%  \
barrels08  37843.0  17.532506  4.57595  0.06  14.33087  17.347895  20.600625   

                 max  
barrels08  47.087143

In [50]:
# option B:
df["barrels08"].describe().to_frame().transpose()

count       mean      std   min       25%        50%        75%  \
barrels08  37843.0  17.532506  4.57595  0.06  14.33087  17.347895  20.600625   

                 max  
barrels08  47.087143

. Iterate over all numerical columns
.  for each column, calculate the IQR
.  Based on the IQR, define the lower and upper threshold for acceptable values
.  identify all outliers
.  append the DataFrame outliers with the rows that were identified as outliers

In [51]:
cols = list(df.columns).append("outlier_column") # we add a new column 'outlier_column' to keep track of the column in which a row was identified as an outlier
outliers = pd.DataFrame(columns=cols)

for col in df.select_dtypes(include=[np.number]).columns:
    descr = df[[col]].describe().transpose()
    iqr = descr.loc[col, "75%"] - descr.loc[col, "25%"]
    cutoff = iqr * 12        # increasing it from 3 in order to cut off less data
    lower_threshold = descr.loc[col, "25%"] - cutoff
    upper_threshold = descr.loc[col, "75%"] + cutoff
    
    results = df.loc[(df[col] < lower_threshold) | (df[col] > upper_threshold)].copy()
    outliers = pd.concat([outliers, results.assign(outlier_column=col)], sort=False)

In [52]:
outliers.shape # a data frame w/rows that#re outliers according to the iqr

(62529, 39)

In [53]:
len(outliers.drop_duplicates())

62529

In [54]:
len(set(outliers.index))   # according to this, the tolerance factor was set

16479

In [55]:
df.shape

(37843, 38)

In [56]:
outliers.index.to_series().value_counts()

29750    11
25673    11
25715    11
25895    11
29443    11
         ..
31972     1
31971     1
31959     1
31958     1
33895     1
Length: 16479, dtype: int64

In [57]:
outliers.loc[29750,] # gives all therows with the same index (29750), the last column says where they come from

barrels08  city08   city08U  co2  co2TailpipeGpm  comb08   comb08U  \
29750       0.18     126  126.4057    0             0.0     114  113.6836   
29750       0.18     126  126.4057    0             0.0     114  113.6836   
29750       0.18     126  126.4057    0             0.0     114  113.6836   
29750       0.18     126  126.4057    0             0.0     114  113.6836   
29750       0.18     126  126.4057    0             0.0     114  113.6836   
29750       0.18     126  126.4057    0             0.0     114  113.6836   
29750       0.18     126  126.4057    0             0.0     114  113.6836   
29750       0.18     126  126.4057    0             0.0     114  113.6836   
29750       0.18     126  126.4057    0             0.0     114  113.6836   
29750       0.18     126  126.4057    0             0.0     114  113.6836   
29750       0.18     126  126.4057    0             0.0     114  113.6836   

       cylinders  displ              drive  ...  pv4           trany  \
29750        0.0    NaN  Front-Wheel Drive  ...    0  Automatic (A1)   
29750        0.0    NaN  Front-Wheel Drive  ...    0  Automatic (A1)   
29750        0.0    NaN  Front-Wheel Drive  ...    0  Automatic (A1)   
29750        0.0    NaN  Front-Wheel Drive  ...    0  Automatic (A1)   
29750        0.0    NaN  Front-Wheel Drive  ...    0  Automatic (A1)   
29750        0.0    NaN  Front-Wheel Drive  ...    0  Automatic (A1)   
29750        0.0    NaN  Front-Wheel Drive  ...    0  Automatic (A1)   
29750        0.0    NaN  Front-Wheel Drive  ...    0  Automatic (A1)   
29750        0.0    NaN  Front-Wheel Drive  ...    0  Automatic (A1)   
29750        0.0    NaN  Front-Wheel Drive  ...    0  Automatic (A1)   
29750        0.0    NaN  Front-Wheel Drive  ...    0  Automatic (A1)   

          UCity  UHighway        VClass  year  youSaveSpend  \
29750  180.5796   144.616  Midsize Cars  2016          3750   
29750  180.5796   144.616  Midsize Cars  2016          3750   
29750  180.5796   144.616  Midsize Cars  2016          3750   
29750  180.5796   144.616  Midsize Cars  2016          3750   
29750  180.5796   144.616  Midsize Cars  2016          3750   
29750  180.5796   144.616  Midsize Cars  2016          3750   
29750  180.5796   144.616  Midsize Cars  2016          3750   
29750  180.5796   144.616  Midsize Cars  2016          3750   
29750  180.5796   144.616  Midsize Cars  2016          3750   
29750  180.5796   144.616  Midsize Cars  2016          3750   
29750  180.5796   144.616  Midsize Cars  2016          3750   

                          createdOn                    modifiedOn  \
29750  Mon Oct 19 00:00:00 EDT 2015  Wed Dec 02 00:00:00 EST 2015   
29750  Mon Oct 19 00:00:00 EDT 2015  Wed Dec 02 00:00:00 EST 2015   
29750  Mon Oct 19 00:00:00 EDT 2015  Wed Dec 02 00:00:00 EST 2015   
29750  Mon Oct 19 00:00:00 EDT 2015  Wed Dec 02 00:00:00 EST 2015   
29750  Mon Oct 19 00:00:00 EDT 2015  Wed Dec 02 00:00:00 EST 2015   
29750  Mon Oct 19 00:00:00 EDT 2015  Wed Dec 02 00:00:00 EST 2015   
29750  Mon Oct 19 00:00:00 EDT 2015  Wed Dec 02 00:00:00 EST 2015   
29750  Mon Oct 19 00:00:00 EDT 2015  Wed Dec 02 00:00:00 EST 2015   
29750  Mon Oct 19 00:00:00 EDT 2015  Wed Dec 02 00:00:00 EST 2015   
29750  Mon Oct 19 00:00:00 EDT 2015  Wed Dec 02 00:00:00 EST 2015   
29750  Mon Oct 19 00:00:00 EDT 2015  Wed Dec 02 00:00:00 EST 2015   

       outlier_column  
29750          city08  
29750         city08U  
29750             co2  
29750          comb08  
29750         comb08U  
29750         feScore  
29750        ghgScore  
29750      highway08U  
29750             hlv  
29750             hpv  
29750           UCity  

[11 rows x 39 columns]

### 4. Convert data types

In [58]:
df.dtypes

barrels08         float64
city08              int64
city08U           float64
co2                 int64
co2TailpipeGpm    float64
comb08              int64
comb08U           float64
cylinders         float64
displ             float64
drive              object
engId               int64
eng_dscr           object
feScore             int64
fuelCost08          int64
fuelType           object
fuelType1          object
ghgScore            int64
highway08           int64
highway08U        float64
hlv                 int64
hpv                 int64
id                  int64
lv2                 int64
lv4                 int64
make               object
model              object
mpgData            object
phevBlended          bool
pv2                 int64
pv4                 int64
trany              object
UCity             float64
UHighway          float64
VClass             object
year                int64
youSaveSpend        int64
createdOn          object
modifiedOn         object
dtype: objec

In [59]:
df["year"] = pd.to_datetime(df["year"], format="%Y") # explicitly pass the string format as year

In [60]:
df.dtypes  # see the year at the end of the list

barrels08                float64
city08                     int64
city08U                  float64
co2                        int64
co2TailpipeGpm           float64
comb08                     int64
comb08U                  float64
cylinders                float64
displ                    float64
drive                     object
engId                      int64
eng_dscr                  object
feScore                    int64
fuelCost08                 int64
fuelType                  object
fuelType1                 object
ghgScore                   int64
highway08                  int64
highway08U               float64
hlv                        int64
hpv                        int64
id                         int64
lv2                        int64
lv4                        int64
make                      object
model                     object
mpgData                   object
phevBlended                 bool
pv2                        int64
pv4                        int64
trany     

### 5. String Cleaning

In [61]:
df["trany"].value_counts() # many of these values are supposed to be the same unique value, one must reduce them

Automatic 4-spd                     11042
Manual 5-spd                         8311
Automatic 3-spd                      3151
Automatic (S6)                       2638
Manual 6-spd                         2429
Automatic 5-spd                      2184
Manual 4-spd                         1483
Automatic 6-spd                      1432
Automatic (S8)                        960
Automatic (S5)                        824
Automatic (variable gear ratios)      681
Automatic 7-spd                       663
Automatic (S7)                        261
Auto(AM-S7)                           256
Automatic 8-spd                       243
Automatic (S4)                        233
Auto(AM7)                             160
Auto(AV-S6)                           145
Auto(AM6)                             110
Automatic (A1)                        109
Auto(AM-S6)                            92
Automatic 9-spd                        90
Manual 3-spd                           77
Manual 7-spd                      

In [62]:
len(df["trany"].value_counts())

46

In [63]:
df["trany"].str.replace("Automatic", "Auto").value_counts() # turn evthng "automatic" to "Auto"

Auto 4-spd                     11042
Manual 5-spd                    8311
Auto 3-spd                      3151
Auto (S6)                       2638
Manual 6-spd                    2429
Auto 5-spd                      2184
Manual 4-spd                    1483
Auto 6-spd                      1432
Auto (S8)                        960
Auto (S5)                        824
Auto (variable gear ratios)      681
Auto 7-spd                       663
Auto (S7)                        261
Auto(AM-S7)                      256
Auto 8-spd                       243
Auto (S4)                        233
Auto(AM7)                        160
Auto(AV-S6)                      145
Auto(AM6)                        110
Auto (A1)                        109
Auto(AM-S6)                       92
Auto 9-spd                        90
Manual 3-spd                      77
Manual 7-spd                      68
Auto(AV-S7)                       63
Auto(AV-S8)                       26
Auto (S9)                         26
M

In [64]:
len(df["trany"].str.replace("Automatic", "Auto").value_counts()) # count after the reassignment

44

In [65]:
len(df["trany"]
    .str.replace("Automatic", "Auto")
    .str.replace("Auto\(", "Auto ")
    .str.replace("Manual\(", "Manual ")
    .str.replace("(", "")
    .str.replace(")", "")
    .str.replace("-", "").value_counts())

/tmp/ipykernel_7857/2405765207.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  len(df["trany"]
/tmp/ipykernel_7857/2405765207.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  len(df["trany"]


38

In [66]:
clean_trany = (df["trany"]            # assigning to a new variable as a series
               .str.replace("Automatic", "Auto")
               .str.replace("Auto\(", "Auto ")
               .str.replace("Manual\(", "Manual ")
               .str.replace("(", "")
               .str.replace(")", "")
               .str.replace("-", ""))

/tmp/ipykernel_7857/3974003119.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  clean_trany = (df["trany"]
/tmp/ipykernel_7857/3974003119.py:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  clean_trany = (df["trany"]


In [67]:
clean_trany.value_counts()

Auto 4spd                    11042
Manual 5spd                   8311
Auto 3spd                     3151
Auto S6                       2638
Manual 6spd                   2429
Auto 5spd                     2184
Manual 4spd                   1483
Auto 6spd                     1433
Auto S8                        960
Auto S5                        824
Auto variable gear ratios      681
Auto 7spd                      663
Auto S7                        261
Auto AMS7                      256
Auto 8spd                      243
Auto S4                        233
Auto AM7                       160
Auto AVS6                      155
Auto AM6                       111
Auto A1                        110
Auto AMS6                       92
Auto 9spd                       90
Manual 3spd                     77
Manual 7spd                     68
Auto AVS7                       63
Auto AVS8                       27
Auto S9                         26
Manual 4spd Doubled             17
Auto AM5            

In [68]:
clean_trany = (clean_trany
               .str.split(" ", expand=True)
               .drop(columns=[2,3])
               .rename(columns={0: "tranType",   # renaming columns
                                1: "gears"}))

In [69]:
clean_trany["gears"] = clean_trany["gears"].str.replace(r"\D", "")

/tmp/ipykernel_7857/3528344760.py:1: FutureWarning: The default value of regex will change from True to False in a future version.
  clean_trany["gears"] = clean_trany["gears"].str.replace(r"\D", "")


In [70]:
clean_trany[clean_trany["gears"] == ""] = pd.NA

In [71]:
clean_trany.isna().sum()    # showing missing values

tranType    698
gears       698
dtype: int64

In [72]:
df = (df
      .assign(tranType=clean_trany["tranType"])
      .assign(gears=clean_trany["gears"])
      .drop(columns="trany"))

In [73]:
df["tranType"].value_counts()

Auto      24756
Manual    12389
Name: tranType, dtype: int64

In [74]:
df["gears"].value_counts() # much cleaner comparing with the beginning

4    12777
5    11334
6     6862
3     3230
7     1474
8     1241
9      117
1      110
Name: gears, dtype: int64

### 6. Drop duplicates

In [75]:
print(f"Before dropping duplicates, our DataFrame had {len(df)} rows") # comparing before and
                                                            # after drop.duplicates
print(f"After dropping duplicates, our DataFrame had {len(df.drop_duplicates())} rows")

Before dropping duplicates, our DataFrame had 37843 rows
After dropping duplicates, our DataFrame had 37843 rows


In [76]:
df.drop_duplicates(["gears"]) # removes rows identical (here there weren't identical rows)

barrels08  city08  city08U  co2  co2TailpipeGpm  comb08  comb08U  \
0      15.695714      19   0.0000   -1      423.190476      21   0.0000   
3      29.964545      10   0.0000   -1      807.909091      11   0.0000   
9      13.184400      23   0.0000   -1      355.480000      25   0.0000   
692    18.311667      15   0.0000   -1      493.722222      18   0.0000   
2903   10.300313      30   0.0000   -1      277.718750      32   0.0000   
10412  19.388824      14   0.0000   -1      522.764706      17   0.0000   
14939  16.480500      17   0.0000   -1      444.350000      20   0.0000   
23026   0.348000      62   0.0000    0        0.000000      58   0.0000   
26790  14.982273      19  19.1415  399      399.000000      22  22.3746   

       cylinders  displ              drive  ...  pv4    UCity  UHighway  \
0            4.0    2.0   Rear-Wheel Drive  ...    0  23.3333   35.0000   
3            8.0    5.2   Rear-Wheel Drive  ...    0  12.2222   16.6667   
9            4.0    1.8  Front-Wheel Drive  ...   89  29.0000   42.0000   
692          8.0    5.7   Rear-Wheel Drive  ...    0  19.0000   32.0000   
2903         4.0    1.6  Front-Wheel Drive  ...    0  39.0000   50.0000   
10412        8.0    5.0   Rear-Wheel Drive  ...    0  17.4000   29.1000   
14939        8.0    4.6   Rear-Wheel Drive  ...  103  20.8489   34.6499   
23026        0.0    NaN      2-Wheel Drive  ...    0  88.6842   76.5909   
26790        6.0    3.2  Front-Wheel Drive  ...    0  24.1000   39.5000   

                                VClass       year youSaveSpend  \
0                          Two Seaters 1985-01-01        -1250   
3                                 Vans 1985-01-01        -8500   
9                         Compact Cars 1993-01-01            0   
692                        Two Seaters 1994-01-01        -4500   
2903                   Subcompact Cars 1996-01-01         1500   
10412                      Two Seaters 2004-01-01        -5250   
14939                     Midsize Cars 2007-01-01        -3250   
23026       Standard Pickup Trucks 2WD 2001-01-01         1000   
26790  Small Sport Utility Vehicle 2WD 2014-01-01         -750   

                          createdOn                    modifiedOn  tranType  \
0      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013    Manual   
3      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013      Auto   
9      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013      Auto   
692    Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013    Manual   
2903   Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013      <NA>   
10412  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013      Auto   
14939  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013      Auto   
23026  Tue Jan 01 00:00:00 EST 2013  Fri Apr 11 00:00:00 EDT 2014      Auto   
26790  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013      Auto   

       gears  
0          5  
3          3  
9          4  
692        6  
2903    <NA>  
10412      7  
14939      8  
23026      1  
26790      9  

[9 rows x 39 columns]

In [77]:
def check_duplicates(df, cols):
    """
    Function to drop duplicates from a df based on a list of columns
    """
    # 1. count no of rows in df before dropping
    # 2. drop duplicate rows based on list of columns
    # 3. count no of rows in df after dropping
    # 4. print no of rows that would be dropped

In [78]:
def check_duplicates(df, cols, return_df=False):
    """
    Function to drop duplicates from a df based on a list of columns
    """
    # 1. count no of rows in df before dropping
    before = len(df)
    
    # 2. drop duplicate rows based on list of columns
    df_after_dropping = df.drop_duplicates(cols)
    
    # 3. count no of rows in df after dropping
    after = len(df_after_dropping)
    
    # 4a. calculate percentage of rows that would be dropped
    perc = int(((before - after) / before) * 100)
    
    # 4b. print no of rows that would be dropped
    if return_df:
        print(f"This list of columns would drop {before - after} rows. {perc}% of all rows.")
        return df_after_dropping
    else:
        print(f"This list of columns would drop {before - after} rows. {perc}% of all rows.")

In [79]:
check_duplicates(df, ["gears"])        # too many rows, so probably not a good idea

This list of columns would drop 37834 rows. 99% of all rows.


In [80]:
columns = ["make", "model", "year", "displ", "cylinders", "tranType", "gears", "drive",
           "VClass", "fuelType", "barrels08", "city08"]

check_duplicates(df, columns)

This list of columns would drop 1355 rows. 3% of all rows.


In [81]:
a = set((1,2,3,3,3,3))
b = set((2,3))

In [82]:
a

{1, 2, 3}

In [83]:
b

{2, 3}

In [84]:
a - b   # subtracting 2 sets

{1}

In [85]:
df_example = df.copy()

In [86]:
df_example = df_example[columns]

In [87]:
df_example_after = df_example.drop_duplicates()

In [88]:
len(df_example)

37843

In [89]:
len(df_example_after)

36488

In [90]:
df_ex_before_index = set(df_example.index)

In [91]:
df_ex_after_index = set(df_example_after.index)

In [92]:
df_ex_dropped_index = df_ex_before_index - df_ex_after_index

In [93]:
len(df_example.loc[df_ex_dropped_index])

/tmp/ipykernel_7857/2304485543.py:1: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.
  len(df_example.loc[df_ex_dropped_index])


1355

In [94]:
df_example.loc[df_ex_dropped_index]

/tmp/ipykernel_7857/1537148038.py:1: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.
  df_example.loc[df_ex_dropped_index]


make                    model       year  displ  cylinders  \
32778    Cadillac  Fleetwood/DeVille (FWD) 1985-01-01    4.3        6.0   
32782     Pontiac                  Firefly 1989-01-01    1.0        3.0   
32784     Pontiac                  Firefly 1989-01-01    1.0        3.0   
16402         GMC           C15 Pickup 2WD 1986-01-01    5.0        8.0   
8222   Mitsubishi                   Mirage 1985-01-01    1.6        4.0   
...           ...                      ...        ...    ...        ...   
24538     Ferrari               California 2012-01-01    4.3        8.0   
24541     Ferrari                       FF 2012-01-01    6.3       12.0   
32744      Jaguar                XJS Coupe 1989-01-01    5.3       12.0   
24565       Honda                   Accord 1987-01-01    2.0        4.0   
32767    Cadillac  Fleetwood/DeVille (FWD) 1985-01-01    4.1        8.0   

      tranType gears                    drive                  VClass  \
32778     Auto     4        Front-Wheel Drive              Large Cars   
32782     Auto     3        Front-Wheel Drive         Subcompact Cars   
32784   Manual     5        Front-Wheel Drive         Subcompact Cars   
16402     Auto     4         Rear-Wheel Drive  Standard Pickup Trucks   
8222      Auto     3        Front-Wheel Drive         Subcompact Cars   
...        ...   ...                      ...                     ...   
24538     Auto     7         Rear-Wheel Drive             Two Seaters   
24541     Auto     7  Part-time 4-Wheel Drive            Midsize Cars   
32744     Auto     3         Rear-Wheel Drive         Subcompact Cars   
24565     Auto     4        Front-Wheel Drive            Compact Cars   
32767     Auto     4        Front-Wheel Drive              Large Cars   

      fuelType  barrels08  city08  
32778   Diesel  16.616739      19  
32782  Regular   9.694412      32  
32784  Regular   8.039268      38  
16402  Regular  21.974000      14  
8222   Premium  14.982273      21  
...        ...        ...     ...  
24538  Premium  21.974000      13  
24541  Premium  25.354615      11  
32744  Regular  25.354615      11  
24565  Regular  14.330870      21  
32767  Regular  18.311667      15  

[1355 rows x 12 columns]

Create a function, that takes a DataFrame and a list of column names and:
1. Calculate the number of rows that would get dropped (check_duplicates()) and tell the user
2. Ask the user, if they want to see the rows that would get dropped
   - 2.1. if yes, return and exit
   - 2.1. if no, ask if user wants to drop the rows
   - 2.1.1 if yes, drop the rows and return new DataFrame
   - 2.1.2. if no, exit

In [95]:
def securely_drop_duplicates(df, cols):
    #Calculate the number of rows that would get dropped (check_duplicates()) and tell the user
    df_after_drop = check_duplicates(df, cols, return_df=True)
    
    #Generate a new df with the rows that would get dropped
    index_before = set(df[cols].index)
    index_after = set(df_after_drop.index)
    index_diff = index_before - index_after
    
    df_dropped_rows = df[cols].loc[index_diff].copy()
    
    #Ask the user, if they want to see the rows that would get dropped
    response1 = input("Before dropping, do you want me to return the rows that would get dropped?")
    
    #2.1. if yes, return and exit
    if response1 == "yes":
        return df_dropped_rows
        
    #2.1. if no, ask if user wants to drop the rows
    elif response1 == "no":
        response2 = input("Do you want me to go ahead and drop the rows?")
    
        #2.1.1 if yes, drop the rows and return new DataFrame
        if response2 == "yes":
            return df_after_drop
        
        #2.1.2. if no, exit
        elif response2 == "no":
            return

In [96]:
input("What's your name?") # just to see if one can use it with "while"

What's your name?Oscar


'Oscar'

In [97]:
def securely_drop_duplicates(df, cols):
    df_after_drop = check_duplicates(df, cols, return_df=True)
    
    index_before = set(df[cols].index)
    index_after = set(df_after_drop.index)
    index_diff = index_before - index_after
    
    df_dropped_rows = df[cols].loc[index_diff].copy()
    
    response1 = None
    while response1 not in ["yes", "no"]:
        response1 = input("Before dropping, do you want me to return the rows that would get dropped? [yes|no]")
    
    if response1 == "yes":
        return df_dropped_rows
    elif response1 == "no":
        
        response2 = input("Do you want me to go ahead and drop the rows?")
    
        if response2 == "yes":
            return df_after_drop
        elif response2 == "no":
            print("Byyyyeeee!")
            return

In [98]:
securely_drop_duplicates(df, columns)

/tmp/ipykernel_7857/3517476651.py:8: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.
  df_dropped_rows = df[cols].loc[index_diff].copy()


This list of columns would drop 1355 rows. 3% of all rows.
Before dropping, do you want me to return the rows that would get dropped? [yes|no]yes


make                    model       year  displ  cylinders  \
32778    Cadillac  Fleetwood/DeVille (FWD) 1985-01-01    4.3        6.0   
32782     Pontiac                  Firefly 1989-01-01    1.0        3.0   
32784     Pontiac                  Firefly 1989-01-01    1.0        3.0   
16402         GMC           C15 Pickup 2WD 1986-01-01    5.0        8.0   
8222   Mitsubishi                   Mirage 1985-01-01    1.6        4.0   
...           ...                      ...        ...    ...        ...   
24538     Ferrari               California 2012-01-01    4.3        8.0   
24541     Ferrari                       FF 2012-01-01    6.3       12.0   
32744      Jaguar                XJS Coupe 1989-01-01    5.3       12.0   
24565       Honda                   Accord 1987-01-01    2.0        4.0   
32767    Cadillac  Fleetwood/DeVille (FWD) 1985-01-01    4.1        8.0   

      tranType gears                    drive                  VClass  \
32778     Auto     4        Front-Wheel Drive              Large Cars   
32782     Auto     3        Front-Wheel Drive         Subcompact Cars   
32784   Manual     5        Front-Wheel Drive         Subcompact Cars   
16402     Auto     4         Rear-Wheel Drive  Standard Pickup Trucks   
8222      Auto     3        Front-Wheel Drive         Subcompact Cars   
...        ...   ...                      ...                     ...   
24538     Auto     7         Rear-Wheel Drive             Two Seaters   
24541     Auto     7  Part-time 4-Wheel Drive            Midsize Cars   
32744     Auto     3         Rear-Wheel Drive         Subcompact Cars   
24565     Auto     4        Front-Wheel Drive            Compact Cars   
32767     Auto     4        Front-Wheel Drive              Large Cars   

      fuelType  barrels08  city08  
32778   Diesel  16.616739      19  
32782  Regular   9.694412      32  
32784  Regular   8.039268      38  
16402  Regular  21.974000      14  
8222   Premium  14.982273      21  
...        ...        ...     ...  
24538  Premium  21.974000      13  
24541  Premium  25.354615      11  
32744  Regular  25.354615      11  
24565  Regular  14.330870      21  
32767  Regular  18.311667      15  

[1355 rows x 12 columns]

In [100]:
securely_drop_duplicates(df, columns)

/tmp/ipykernel_7857/3517476651.py:8: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.
  df_dropped_rows = df[cols].loc[index_diff].copy()


This list of columns would drop 1355 rows. 3% of all rows.
Before dropping, do you want me to return the rows that would get dropped? [yes|no]no
Do you want me to go ahead and drop the rows?no
Byyyyeeee!


In [101]:
dupl = securely_drop_duplicates(df, columns)

/tmp/ipykernel_7857/3517476651.py:8: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.
  df_dropped_rows = df[cols].loc[index_diff].copy()


This list of columns would drop 1355 rows. 3% of all rows.
Before dropping, do you want me to return the rows that would get dropped? [yes|no]yes


In [102]:
dupl

make                    model       year  displ  cylinders  \
32778    Cadillac  Fleetwood/DeVille (FWD) 1985-01-01    4.3        6.0   
32782     Pontiac                  Firefly 1989-01-01    1.0        3.0   
32784     Pontiac                  Firefly 1989-01-01    1.0        3.0   
16402         GMC           C15 Pickup 2WD 1986-01-01    5.0        8.0   
8222   Mitsubishi                   Mirage 1985-01-01    1.6        4.0   
...           ...                      ...        ...    ...        ...   
24538     Ferrari               California 2012-01-01    4.3        8.0   
24541     Ferrari                       FF 2012-01-01    6.3       12.0   
32744      Jaguar                XJS Coupe 1989-01-01    5.3       12.0   
24565       Honda                   Accord 1987-01-01    2.0        4.0   
32767    Cadillac  Fleetwood/DeVille (FWD) 1985-01-01    4.1        8.0   

      tranType gears                    drive                  VClass  \
32778     Auto     4        Front-Wheel Drive              Large Cars   
32782     Auto     3        Front-Wheel Drive         Subcompact Cars   
32784   Manual     5        Front-Wheel Drive         Subcompact Cars   
16402     Auto     4         Rear-Wheel Drive  Standard Pickup Trucks   
8222      Auto     3        Front-Wheel Drive         Subcompact Cars   
...        ...   ...                      ...                     ...   
24538     Auto     7         Rear-Wheel Drive             Two Seaters   
24541     Auto     7  Part-time 4-Wheel Drive            Midsize Cars   
32744     Auto     3         Rear-Wheel Drive         Subcompact Cars   
24565     Auto     4        Front-Wheel Drive            Compact Cars   
32767     Auto     4        Front-Wheel Drive              Large Cars   

      fuelType  barrels08  city08  
32778   Diesel  16.616739      19  
32782  Regular   9.694412      32  
32784  Regular   8.039268      38  
16402  Regular  21.974000      14  
8222   Premium  14.982273      21  
...        ...        ...     ...  
24538  Premium  21.974000      13  
24541  Premium  25.354615      11  
32744  Regular  25.354615      11  
24565  Regular  14.330870      21  
32767  Regular  18.311667      15  

[1355 rows x 12 columns]

In [103]:
df_new = securely_drop_duplicates(df, columns)

/tmp/ipykernel_7857/3517476651.py:8: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.
  df_dropped_rows = df[cols].loc[index_diff].copy()


This list of columns would drop 1355 rows. 3% of all rows.
Before dropping, do you want me to return the rows that would get dropped? [yes|no]no
Do you want me to go ahead and drop the rows?yes


In [104]:
df_new

barrels08  city08  city08U  co2  co2TailpipeGpm  comb08  comb08U  \
0      15.695714      19      0.0   -1      423.190476      21      0.0   
1      29.964545       9      0.0   -1      807.909091      11      0.0   
2      12.207778      23      0.0   -1      329.148148      27      0.0   
3      29.964545      10      0.0   -1      807.909091      11      0.0   
4      17.347895      17      0.0   -1      467.736842      19      0.0   
...          ...     ...      ...  ...             ...     ...      ...   
37838  14.982273      19      0.0   -1      403.954545      22      0.0   
37839  14.330870      20      0.0   -1      386.391304      23      0.0   
37840  15.695714      18      0.0   -1      423.190476      21      0.0   
37841  15.695714      18      0.0   -1      423.190476      21      0.0   
37842  18.311667      16      0.0   -1      493.722222      18      0.0   

       cylinders  displ                       drive  ...  pv4    UCity  \
0            4.0    2.0            Rear-Wheel Drive  ...    0  23.3333   
1           12.0    4.9            Rear-Wheel Drive  ...    0  11.0000   
2            4.0    2.2           Front-Wheel Drive  ...    0  29.0000   
3            8.0    5.2            Rear-Wheel Drive  ...    0  12.2222   
4            4.0    2.2  4-Wheel or All-Wheel Drive  ...   90  21.0000   
...          ...    ...                         ...  ...  ...      ...   
37838        4.0    2.2           Front-Wheel Drive  ...   90  24.0000   
37839        4.0    2.2           Front-Wheel Drive  ...   90  25.0000   
37840        4.0    2.2  4-Wheel or All-Wheel Drive  ...   90  23.0000   
37841        4.0    2.2  4-Wheel or All-Wheel Drive  ...   90  23.0000   
37842        4.0    2.2  4-Wheel or All-Wheel Drive  ...   90  20.0000   

       UHighway           VClass       year youSaveSpend  \
0       35.0000      Two Seaters 1985-01-01        -1250   
1       19.0000      Two Seaters 1985-01-01        -8500   
2       47.0000  Subcompact Cars 1985-01-01          500   
3       16.6667             Vans 1985-01-01        -8500   
4       32.0000     Compact Cars 1993-01-01        -4000   
...         ...              ...        ...          ...   
37838   37.0000     Compact Cars 1993-01-01         -750   
37839   39.0000     Compact Cars 1993-01-01         -500   
37840   34.0000     Compact Cars 1993-01-01        -1250   
37841   34.0000     Compact Cars 1993-01-01        -1250   
37842   29.0000     Compact Cars 1993-01-01        -4500   

                          createdOn                    modifiedOn  tranType  \
0      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013    Manual   
1      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013    Manual   
2      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013    Manual   
3      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013      Auto   
4      Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013    Manual   
...                             ...                           ...       ...   
37838  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013      Auto   
37839  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013    Manual   
37840  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013      Auto   
37841  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013    Manual   
37842  Tue Jan 01 00:00:00 EST 2013  Tue Jan 01 00:00:00 EST 2013      Auto   

       gears  
0          5  
1          5  
2          5  
3          3  
4          5  
...      ...  
37838      4  
37839      5  
37840      4  
37841      5  
37842      4  

[36488 rows x 39 columns]